SMT back-end: Extend trace value evaluation

tautschnig opened the pull request:
We record the values of expressions by adding `expr == expr` as constraints in order to be able to fetch and display them in the GOTO trace (for declaration; we presently also introduce new symbols in other cases). Constructing a trace then requires the ability to evaluate all kinds of expressions to obtain the value for those expressions.

Fixes: #7308

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
SMT back-end: Extend trace value evaluation

We record the values of expressions by adding `expr == expr`
as constraints in order to be able to fetch and display them in the GOTO
trace (for declaration; we presently also introduce new symbols in other
cases). Constructing a trace then requires the ability to evaluate all
kinds of expressions to obtain the value for those expressions.

Fixes: #7308

TGWDB commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -321,16 +313,23 @@ exprt smt2_convt::get(const exprt &expr) const
     {
-      element = get(element);
+      exprt eval_op = get(op);
+      if(eval_op.is_nil())
+        return nil_exprt{};
```
I don't quite understand why we return nil if any operand could be nil? The old code here (e.g.) for arrays would leave the nil in place and return the larger expr.

tautschnig replied:
I'd fear that we fail to notice problems. I'm not sure returning `nil` ever is the right thing to do?

kroening commented:
The question is whether the decision procedure or the user of the decision procedure ought to do the simplification.

peterschrammel reviewed (commented):
> `Reason: full_lhs_value in assignment must not be nil`

How does simplification make it non-nil?

tautschnig commented:
@kroening said:
> The question is whether the decision procedure or the user of the decision procedure ought to do the simplification.

I accept that the changes introducing simplification to io_args could be factored out, they were a bit of a drive-by cleanup. We already do this in other branches of `build_goto_trace`, so it just seemed like a good idea to make code simpler. Let me know if you want me to remove those from the PR.

@peterschrammel said:
>> `Reason: full_lhs_value in assignment must not be nil`

> How does simplification make it non-nil?

Would you mind elaborating for I'm not sure how this comment came about?

peterschrammel commented:
> Would you mind elaborating for I'm not sure how this comment came about?

I'm referring to the [issue](https://github.com/diffblue/cbmc/issues/7308) that is supposed to fixed here. I'm not entirely sure I understand how simplification fixes it.

tautschnig commented:
> > Would you mind elaborating for I'm not sure how this comment came about?
> 
> I'm referring to the [issue](https://github.com/diffblue/cbmc/issues/7308) that is supposed to fixed here. I'm not entirely sure I understand how simplification fixes it.

Ah! The issue is fixed by the changes in this PR as https://github.com/diffblue/cbmc/pull/7310/files#diff-0eb7414a51aacc433e17eec049af7ed61dd62d3ee3c4b466f991c60a861c7fccR390-R391 will now receive a non-nil value. Not a matter of simplification at all.

tautschnig commented:
> The question is whether the decision procedure or the user of the decision procedure ought to do the simplification.

I have removed the simplification changes from this PR and moved them to #7323 instead.
